test_util/onprem.py

```diff
@@ -0,0 +1,203 @@
+        log.error(response.content.decode())
+        response.raise_for_status()
+
+
+class OnpremCluster:
+
+    def __init__(self, ssher, masters, private_agents, public_agents, bootstrap_host):
+        """ input lists are lists of Host Tuples
+        """
+        self.ssher = ssher
+        self.masters = masters
+        self.private_agents = private_agents
+        self.public_agents = public_agents
+        self.bootstrap_host = bootstrap_host
+        assert all(h.private_ip for h in self.hosts), (
+            'All cluster hosts require a private IP. hosts: {}'.format(repr(self.hosts))
+        )
+
+    def get_master_ips(self):
+        return copy.copy(self.masters)
+
+    def get_private_agent_ips(self):
+        return copy.copy(self.private_agents)
+
+    def get_public_agent_ips(self):
+        return copy.copy(self.public_agents)
+
+    @classmethod
+    def from_hosts(cls, ssher, hosts, num_masters, num_private_agents, num_public_agents):
+        bootstrap_host, masters, private_agents, public_agents = (
+            cls.partition_cluster(hosts, num_masters, num_private_agents, num_public_agents))
+
+        return cls(
+            ssher=ssher,
+            masters=masters,
+            private_agents=private_agents,
+            public_agents=public_agents,
+            bootstrap_host=bootstrap_host,
+        )
+
+    @property
+    def hosts(self):
+        return self.masters + self.private_agents + self.public_agents + (
+            [self.bootstrap_host] if self.bootstrap_host else []
+        )
+
+    @staticmethod
+    def partition_cluster(hosts, num_masters, num_agents, num_public_agents):
+        """Return (bootstrap, masters, agents, public_agents) from hosts."""
+        hosts_iter = iter(sorted(hosts))
+        return (
+            next(hosts_iter),
+            list(itertools.islice(hosts_iter, num_masters)),
+            list(itertools.islice(hosts_iter, num_agents)),
+            list(itertools.islice(hosts_iter, num_public_agents)),
+        )
+
+    def start_bootstrap_zk(self):
+        zk_host = self.bootstrap_host.private_ip + ':2181'
+        self.ssher.command(
+            self.bootsrap_host.private_ip,
+            ['docker', 'run', '--name', 'dcos-bootstrap-zk', '--detach=true',
+             '--publish=2181:2181', '--publish=2888:2888', '--publish=3888:3888', 'jplock/zookeeper'])
+        return zk_host
+
+    def setup_installer_server(self, installer_url: str, offline_mode: bool):
+        log.info('Setting up installer on bootstrap host')
+        return DcosInstallerApiSession.api_session_from_host(
+            self.ssher, self.bootstrap_host.public_ip, installer_url, offline_mode)
+
+
+@retry(wait_fixed=3000, stop_max_delay=300 * 1000)
+def download_dcos_installer(ssher, host, installer_path, download_url):
+    """Response status 403 is fatal for curl's retry. Additionally, S3 buckets
+    have been returning 403 for valid uploads for 10-15 minutes after CI finished build
```

Comment:
:(

If this is need not be exposed, you could make this a private.
Else, please consider documenting and typing the args.